File: src/demos/demoParte26.js
const crypto = require("crypto");

// demo que muestra como cifrar y descifrar con aes gcm usando la libreria crypto de nodejs mostrando el nonce el tag y el texto cifrado en hex
function cifrarAesGcm(plainText, clave) {
  const nonce = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", clave, nonce);
  const cipherText = Buffer.concat([cipher.update(plainText), cipher.final()]);
  const tag = cipher.getAuthTag();

  return { nonce, ciphertext: cipherText, tag };
}

function descifrarAesGcm(nonce, ciphertext, tag, clave) {
  const decipher = crypto.createDecipheriv("aes-256-gcm", clave, nonce);
  decipher.setAuthTag(tag);

  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

File: src/demos/demoParte26.py
from Crypto.Cipher import AES
import secrets


# demo que muestra cifrar y decifrar usando AES em modo gcm usando
def cifrarAesGcm(plaintext: bytes, clave: bytes) -> tuple[bytes, bytes, bytes]:
    nonce = secrets.token_bytes(12)
    cipher = AES.new(clave, AES.MODE_GCM, nonce=nonce)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)
    return nonce, ciphertext, tag


def descifrarAesGcm(nonce: bytes, ciphertext: bytes, tag: bytes, clave: bytes) -> bytes:
    cipher = AES.new(clave, AES.MODE_GCM, nonce=nonce)
    return cipher.decrypt_and_verify(ciphertext, tag)
